File: web/src/types/task.ts
export type TaskStatus =
    | 'created'
    | 'queued'
    | 'claimed'
    | 'in_progress'
    | 'awaiting_review'
    | 'pending_rework'
    | 'completed'
    | 'cancelled'
    | 'failed';

export type TaskPriority = 'low' | 'normal' | 'high' | 'urgent';

export type IterationStatus = 'success' | 'failed' | 'cancelled';

export type FeedbackType = 'approve' | 'request_changes' | 'abort';

export interface AgentMessage {
    timestamp: string;
    role: string;
    content: string;
}

export interface IterationResult {
    status: IterationStatus;
    summary: string;
    files_changed: string[];
    commits: string[];
    agent_messages: AgentMessage[];
}

export interface HumanFeedback {
    provided_at: string;
    feedback_type: FeedbackType;
    message: string;
}

export interface TaskIteration {
    iteration_id: number;
    started_at: string;
    completed_at: string | null;
    result: IterationResult | null;
    human_feedback: HumanFeedback | null;
}

export interface Task {
    id: string;
    repo_url: string;
    description: string;
    base_branch: string;
    target_branch: string;
    status: TaskStatus;
    priority: TaskPriority;
    created_at: string;
    updated_at: string;
    claimed_by: string | null;
}

export interface CreateTaskRequest {
    repo_url: string;
    description: string;
    base_branch?: string;
    target_branch?: string;
    priority?: TaskPriority;
}

export interface CreateTaskResponse {
    task_id: string;
}

export interface ListTasksQuery {
    status?: TaskStatus;
    limit?: number;
    offset?: number;
}

export interface TaskListResponse {
    tasks: Task[];
    total: number;
}

export interface SubmitFeedbackRequest {
    feedback_type: FeedbackType;
    message: string;
}

export interface ReviewData {
    messages: AgentMessage[];
    diff: string;
}

File: src/protocol/task.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TaskStatus {
    Created,
    Queued,
    Claimed,
    InProgress,
    AwaitingReview,
    Iterating,
    Completed,
    Cancelled,
}

impl TaskStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            TaskStatus::Created => "created",
            TaskStatus::Queued => "queued",
            TaskStatus::Claimed => "claimed",
            TaskStatus::InProgress => "in_progress",
            TaskStatus::AwaitingReview => "awaiting_review",
            TaskStatus::Iterating => "iterating",
            TaskStatus::Completed => "completed",
            TaskStatus::Cancelled => "cancelled",
        }
    }

    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "created" => Some(TaskStatus::Created),
            "queued" => Some(TaskStatus::Queued),
            "claimed" => Some(TaskStatus::Claimed),
            "in_progress" => Some(TaskStatus::InProgress),
            "awaiting_review" => Some(TaskStatus::AwaitingReview),
            "iterating" => Some(TaskStatus::Iterating),
            "completed" => Some(TaskStatus::Completed),
            "cancelled" => Some(TaskStatus::Cancelled),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum TaskPriority {
    Low = 0,
    Normal = 1,
    High = 2,
    Urgent = 3,
}

impl TaskPriority {
    pub fn as_i32(&self) -> i32 {
        *self as i32
    }

    pub fn from_i32(v: i32) -> Option<Self> {
        match v {
            0 => Some(TaskPriority::Low),
            1 => Some(TaskPriority::Normal),
            2 => Some(TaskPriority::High),
            3 => Some(TaskPriority::Urgent),
            _ => None,
        }
    }
}

impl Default for TaskPriority {
    fn default() -> Self {
        TaskPriority::Normal
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum IterationStatus {
    Success,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FeedbackType {
    Approve,
    RequestChanges,
    Abort,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Task {
    pub id: Uuid,
    pub repo_url: String,
    pub description: String,
    pub base_branch: String,
    pub target_branch: String,
    pub status: TaskStatus,
    pub priority: TaskPriority,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub claimed_by: Option<Uuid>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentMessage {
    pub timestamp: DateTime<Utc>,
    pub role: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HumanFeedback {
    pub provided_at: DateTime<Utc>,
    pub feedback_type: FeedbackType,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateTaskRequest {
    pub repo_url: String,
    pub description: String,
    pub base_branch: Option<String>,
    pub target_branch: Option<String>,
    pub priority: Option<TaskPriority>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateTaskResponse {
    pub task_id: Uuid,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListTasksQuery {
    pub status: Option<TaskStatus>,
    pub limit: Option<u32>,
    pub offset: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskListResponse {
    pub tasks: Vec<Task>,
    pub total: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubmitFeedbackRequest {
    pub feedback_type: FeedbackType,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClaimTaskRequest {
    pub worker_id: Uuid,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClaimTaskResponse {
    pub task: Option<Task>,
}
